source/hp/refinement.cc

```diff
@@ -635,6 +636,28 @@ namespace hp
     void
     choose_p_over_h(const hp::DoFHandler<dim, spacedim> &dof_handler)
     {
+      // Siblings of cells to be coarsened may not be owned by the same
+      // processor. We will exchange coarsening flags on ghost cells and
+      // temporarily store them.
+      std::map<CellId, std::pair<bool, bool>> ghost_buffer;
+
+      auto pack = [](const typename dealii::hp::DoFHandler<dim, spacedim>::
+                       active_cell_iterator &cell) -> std::pair<bool, bool> {
+        return {cell->coarsen_flag_set(), cell->future_fe_index_set()};
+      };
+
+      auto unpack =
+        [&ghost_buffer](const typename dealii::hp::DoFHandler<dim, spacedim>::
+                          active_cell_iterator &    cell,
+                        const std::pair<bool, bool> pair) -> void {
+        ghost_buffer.emplace(cell->id(), pair);
+      };
+
+      GridTools::exchange_cell_data_to_ghosts<
+        std::pair<bool, bool>,
+        dealii::hp::DoFHandler<dim, spacedim>>(dof_handler, pack, unpack);
```

Comment:
This is just so elegant!

Reply:
I agree. This way to communicate data is indeed elegant.

However, in this particular scenario, we certainly communicate more than necessary. But since this is not critical for performance, I'll leave it like this. Maybe my use of the expression 'elegant' in the pull request message was not the proper one - I should've used 'efficient' in this context. :)